File: cpp/Ice/callback/SimpleWakeUpService.cpp
// Copyright (c) ZeroC, Inc.

#include "SimpleWakeUpService.h"

#include <algorithm>
#include <iostream>

using namespace EarlyRiser;
using namespace std;

namespace
{
    // Converts a time stamp to a time point.
    chrono::system_clock::time_point toTimePoint(int64_t timeStamp)
    {
        const int daysBeforeEpoch = 719162;

        chrono::microseconds timePointMicro{timeStamp / 10}; // timeStamp is in ticks (100 nanoseconds)
        return chrono::system_clock::time_point{timePointMicro - daysBeforeEpoch * 24h};
    }
}

Server::SimpleWakeUpService::~SimpleWakeUpService()
{
    // Waits for all outstanding tasks to complete.
    for (auto& task : _tasks)
    {
        task.wait();
    }
}

void
Server::SimpleWakeUpService::wakeMeUp(optional<AlarmClockPrx> alarmClock, int64_t timeStamp, const Ice::Current&)
{
    if (!alarmClock)
    {
        // Should never happen, but in case it does, the Ice runtime will send an Ice::UnknownException to the client.
        throw std::invalid_argument{"alarmClock is null"};
    }

    // With C++20, we'll be able to print the time point easily--but not with C++17.
    cout << "Dispatching wakeMeUp request { alarmClock = '" << alarmClock << "', timeStamp = " << timeStamp
         << " ticks }" << endl;

    chrono::system_clock::time_point timePoint = toTimePoint(timeStamp);

    // Schedule a wake-up call in a background task.
    _tasks.emplace_back(std::async(
        std::launch::async,
        [alarmClock = std::move(alarmClock), timePoint]()
        {
            // Sleep until the specified time point.
            this_thread::sleep_until(timePoint);

            // First ring.
            ButtonPressed buttonPressed = alarmClock->ring("It's time to wake up!");

            // Keep ringing every 10 seconds until the user presses the stop button.
            while (buttonPressed == ButtonPressed::Snooze)
            {
                this_thread::sleep_for(10s);
                buttonPressed = alarmClock->ring("No more snoozing!");
            }
        }));

    // We don't want the _tasks vector to grow forever so remove all completed tasks here, without waiting.
    // TODO: switch to std::erase_if when we can use C++20.
    _tasks.erase(
        std::remove_if(
            _tasks.begin(),
            _tasks.end(),
            [](const auto& task) { return task.wait_for(std::chrono::seconds(0)) == std::future_status::ready; }),
        _tasks.end());
}

File: python/Glacier2/callback/server/simple_wake_up_service.py
#!/usr/bin/env python
# Copyright (c) ZeroC, Inc.

import asyncio
import time
import Ice
from EarlyRiser import ButtonPressed, WakeUpService, AlarmClockPrx
from datetime import datetime
from common.time import toDatetime

async def callRing(alarmClock: AlarmClockPrx, wakeUpDatetime: datetime):
    """
    Asynchronously call the ring method on the alarm clock proxy until the user presses the stop button.
    The first ring is delayed until the specified wakeUpDatetime. Then the method will keep ringing every 10
    seconds until the user presses the stop button.

    Parameters
    ----------
    alarmClock : AlarmClockPrx
        The proxy to the client's alarm clock.
    wakeUpDatetime : datetime
        The datetime when the alarm should ring.
    """
    delay = wakeUpDatetime.timestamp() - time.time()
    if delay > 0:
        await asyncio.sleep(delay)

    # First ring. This invocation reuses the connection established by the client.
    buttonPressed = await alarmClock.ringAsync("It's time to wake up!")

    # Keep ringing every 10 seconds until the user presses the stop button.
    while buttonPressed is ButtonPressed.Snooze:
        await asyncio.sleep(10)
        buttonPressed = await alarmClock.ringAsync("No more snoozing!")


class SimpleWakeUpService(WakeUpService):
    """
    SimpleWakeUpService is an Ice servant that implements Slice interface WakeUpService.
    """

    async def wakeMeUp(self, alarmClock: AlarmClockPrx, timestamp: int, current: Ice.Current) -> None:
        """
        Implements the abstract method wakeMeUp from the WakeUpService class generated by the Slice compiler.

        Parameters
        ----------
        alarmClock : AlarmClockPrx
            The proxy to the client's alarm clock.
        timestamp : int
            The timestamp represented as the number of ticks since January 1, 0001 00:00:00 UTC.
            Each tick is 100 nanoseconds.
        current : Ice.Current
            The Current object for the dispatch.
        """
        wakeUpDatetime = toDatetime(timestamp)
        print(f"Dispatching wakeMeUp request {{ timeStamp = '{wakeUpDatetime.astimezone()}' }}")

        # Schedule the callRing coroutine to run in the asyncio event loop. Since Ice dispatches async methods in the
        # configured event loop, we can use asyncio.get_running_loop to retrieve it.
        asyncio.get_running_loop().create_task(callRing(alarmClock, wakeUpDatetime))
        print("Client pressed Stop on alarm clock.")
